exonum/src/node/connect_list.rs

```diff
@@ -45,31 +67,68 @@ impl ConnectList {
     }
 
     /// Check if we allow to connect to `address`.
-    pub fn is_address_allowed(&self, address: &SocketAddr) -> bool {
-        self.peers.values().any(|a| a == address)
+    pub fn is_address_allowed(&self, address: &str) -> bool {
+        self.peers.values().any(|a| a.address == address)
     }
 
     /// Adds peer to the ConnectList.
     pub fn add(&mut self, peer: ConnectInfo) {
-        self.peers.insert(peer.public_key, peer.address);
+        self.peers
+            .insert(peer.public_key, PeerAddress::new(peer.address));
+    }
+
+    /// Get public key corresponding to validator with `address`.
+    pub fn find_key_by_resolved_address(&self, address: &SocketAddr) -> Option<&PublicKey> {
+        self.peers
+            .iter()
+            .find(|(_, a)| a.resolved.as_ref() == Some(address))
+            .map(|(p, _)| p)
     }
 
     /// Get public key corresponding to validator with `address`.
-    pub fn find_key_by_address(&self, address: &SocketAddr) -> Option<&PublicKey> {
+    pub fn find_key_by_unresolved_address(&self, address: &str) -> Option<&PublicKey> {
         self.peers
             .iter()
-            .find(|(_, a)| a == &address)
+            .find(|(_, a)| a.address.as_str() == address)
             .map(|(p, _)| p)
     }
+
+    /// Resolves network address and stores it in the `ConnectList`.
+    pub fn resolve_and_cache_peer_address(&mut self, address: &str) -> Option<SocketAddr> {
+        let key = *self.find_key_by_unresolved_address(address)?;
+        let entry = self.peers.get_mut(&key).unwrap();
+
+        let resolved_vec: Vec<SocketAddr> = entry
+            .address
+            .to_socket_addrs()
+            .map(|i| i.collect())
+            .unwrap_or_default();
+
+        entry.resolved_cache.retain(|a| resolved_vec.contains(a));
+
+        if entry.resolved_cache.is_empty() {
+            entry.resolved_cache = resolved_vec;
+        }
+
+        let resolved = entry.resolved_cache.pop();
```

Comment:
Why we need `pop` here?

Reply:
It is used for removing address that we have tried already(in case hostname was resolved to multiple IPs)